### PR TITLE
ODataDBSearch

### PR DESCRIPTION
**Pull request by cristinamanum**
Add OData search queries that will reach DB.

### CODE REVIEW

File: src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs [hunk 1]
```diff
@@ -156,6 +156,30 @@ public async Task UnlistAsync(string id, string version, ITestOutputHelper logge
...
+                request.Headers.Add(ApiKeyHeader, _testSettings.ApiKey);
+
+                using (var response = await httpClient.SendAsync(request))
+                {
+                    await response.EnsureSuccessStatusCodeOrLogAsync(url, logger);
+                    var responseContent = response.Content != null ? await response.Content.ReadAsStringAsync() : string.Empty; 
+                    logger.WriteLine($"The response to the request {url} is {responseContent}");
+                    if(!responseContent.Contains(id))
+                    {
+                        throw new Exception("The expected package id was not found in db.");
```
Comment (zhhyu) [inline src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs]: Is that better if the exception message contains the package id?

File: src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs [hunk 2]
```diff
@@ -156,6 +156,30 @@ public async Task UnlistAsync(string id, string version, ITestOutputHelper logge
             await SendToPackageAsync(HttpMethod.Delete, id, version, logger);
         }
 
+        public async Task SearchPackageODataV2FromDBAsync(string id, bool semver2, ITestOutputHelper logger)
+        {
+            var galleryEndpoint = await GetGalleryUrlAsync(logger);
+            var url = semver2 
+                ? $"{galleryEndpoint}/api/v2/Packages()?$filter=tolower(Id) eq '{id.ToLower()}'&$orderby=Id&semVerLevel=2.0.0"
```
Comment (joelverhagen) [inline src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs]: `eq` is case insensitive on non-hijacked queries.

Comment (joelverhagen) [reply src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs]: If we're trying to break hijacking, `and 1 eq 1` is an easy way

File: src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs [hunk 3]
```diff
@@ -156,6 +156,30 @@ public async Task UnlistAsync(string id, string version, ITestOutputHelper logge
             await SendToPackageAsync(HttpMethod.Delete, id, version, logger);
         }
 
+        public async Task SearchPackageODataV2FromDBAsync(string id, bool semver2, ITestOutputHelper logger)
```
Comment (joelverhagen) [inline src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs]: Why is this not version specific?

Why is `semver2` an option here? Can't we just always use `semVerLevel=2.0.0`

File: src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs [hunk 4]
```diff
@@ -156,6 +156,30 @@ public async Task UnlistAsync(string id, string version, ITestOutputHelper logge
...
+        {
+            var galleryEndpoint = await GetGalleryUrlAsync(logger);
+            var url = semver2 
+                ? $"{galleryEndpoint}/api/v2/Packages()?$filter=tolower(Id) eq '{id.ToLower()}'&$orderby=Id&semVerLevel=2.0.0"
+                : $"{galleryEndpoint}/api/v2/Packages()?$filter=tolower(Id) eq '{id.ToLower()}'&$orderby=Id";
+            using (var httpClient = new HttpClient())
+            using (var request = new HttpRequestMessage(HttpMethod.Get, url))
+            {
+                request.Headers.Add(ApiKeyHeader, _testSettings.ApiKey);
```
Comment (joelverhagen) [inline src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs]: Why send API key?

File: src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs [hunk 5]
```diff
@@ -156,6 +156,30 @@ public async Task UnlistAsync(string id, string version, ITestOutputHelper logge
...
+
+                using (var response = await httpClient.SendAsync(request))
+                {
+                    await response.EnsureSuccessStatusCodeOrLogAsync(url, logger);
+                    var responseContent = response.Content != null ? await response.Content.ReadAsStringAsync() : string.Empty;
```
Comment (joelverhagen) [inline src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs]: Response will never be null

File: src/NuGet.Services.EndToEnd/PushTests.cs [hunk 6]
```diff
@@ -39,6 +39,7 @@ public async Task NewlyPushedIsAvailableInV3(PackageType packageType, bool semVe
             await _clients.FlatContainer.WaitForPackageAsync(package.Id, package.NormalizedVersion, _logger);
             await _clients.Registration.WaitForPackageAsync(package.Id, package.FullVersion, semVer2, logger: _logger);
             await _clients.V2V3Search.WaitForPackageAsync(package.Id, package.FullVersion, _logger);
+            await _clients.Gallery.SearchPackageODataV2FromDBAsync(package.Id, semVer2, _logger);
```
Comment (joelverhagen) [inline src/NuGet.Services.EndToEnd/PushTests.cs]: Why is this assertion in a `V3` test?

File: src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs (hunk 1 above)
Comment (joelverhagen) [reply src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs]: Please throw something more specific that `Exception` like `InvalidOperationException` or just `Assert.Contains`

Comment (cristinamanum) [reply src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs]: Updated to Assert.Contains(id, responseContent)

File: src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs (hunk 2 above)
Comment (cristinamanum) [reply src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs]: Updated the query

File: src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs (hunk 3 above)
Comment (cristinamanum) [reply src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs]: I removed the semVerLevel.

File: src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs (hunk 4 above)
Comment (cristinamanum) [reply src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs]: Not needed - removed.

File: src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs (hunk 5 above)
Comment (cristinamanum) [reply src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs]: Updated

File: src/NuGet.Services.EndToEnd/PushTests.cs (hunk 6 above)
Comment (cristinamanum) [reply src/NuGet.Services.EndToEnd/PushTests.cs]: Moved to a different test.

File: src/NuGet.Services.EndToEnd/PushTests.cs [hunk 7]
```diff
@@ -91,5 +91,20 @@ public async Task NewSemVer2PackageIsFiltered(PackageType packageType)
                 Assert.Equal(1, shouldNotBeEmptyAutocomplete.Data.Count);
             }
         }
+
+        /// <summary>
+        /// Push a package to the gallery and search in OData with non-hijacked query.
+        /// </summary>
+        [Theory]
+        [InlineData(PackageType.SemVer1Stable, false)]
+        public async Task NewlyPushedIsODataSearchableInDB(PackageType packageType, bool semVer2)
```
Comment (joelverhagen) [inline src/NuGet.Services.EndToEnd/PushTests.cs]: `semVer2` is unused

File: src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs [hunk 8]
```diff
@@ -156,6 +157,23 @@ public async Task UnlistAsync(string id, string version, ITestOutputHelper logge
             await SendToPackageAsync(HttpMethod.Delete, id, version, logger);
         }
 
+        public async Task SearchPackageODataV2FromDBAsync(string id, string normalizedVersion, ITestOutputHelper logger)
+        {
+            var galleryEndpoint = await GetGalleryUrlAsync(logger);
+            var url = $"{galleryEndpoint}/api/v2/Packages()?$filter=tolower(Id) eq '{id.ToLower()}' and NormalizedVersion eq '{normalizedVersion}' and 1 eq 1";
```
Comment (joelverhagen) [inline src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs]: you don't need the `tolower`. `Id eq '{id}'` is enough.

Query strings should be built with proper encoding. See `AppendAutocompletePackageIdsQueryString` below and the use of `QueryHelpers.AddQueryString`

File: src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs [hunk 9]
```diff
@@ -156,6 +157,23 @@ public async Task UnlistAsync(string id, string version, ITestOutputHelper logge
...
+            using (var httpClient = new HttpClient())
+            using (var request = new HttpRequestMessage(HttpMethod.Get, url))
+            {
+                using (var response = await httpClient.SendAsync(request))
+                {
+                    await response.EnsureSuccessStatusCodeOrLogAsync(url, logger);
+                    var responseContent = await response.Content.ReadAsStringAsync(); 
+                    logger.WriteLine($"The response to the request {url} is {responseContent}");
```
Comment (joelverhagen) [inline src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs]: `EnsureSuccessStatusCodeOrLogAsync` already logs the response

File: src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs [hunk 10]
```diff
@@ -156,6 +157,23 @@ public async Task UnlistAsync(string id, string version, ITestOutputHelper logge
             await SendToPackageAsync(HttpMethod.Delete, id, version, logger);
         }
 
+        public async Task SearchPackageODataV2FromDBAsync(string id, string normalizedVersion, ITestOutputHelper logger)
+        {
+            var galleryEndpoint = await GetGalleryUrlAsync(logger);
+            var url = $"{galleryEndpoint}/api/v2/Packages()?$filter=tolower(Id) eq '{id.ToLower()}' and NormalizedVersion eq '{normalizedVersion}' and 1 eq 1";
+            using (var httpClient = new HttpClient())
```
Comment (loic-sharma) [inline src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs]: Can you use `_httpClient ` instead?

File: src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs [hunk 11]
```diff
@@ -156,6 +157,23 @@ public async Task UnlistAsync(string id, string version, ITestOutputHelper logge
...
+            using (var request = new HttpRequestMessage(HttpMethod.Get, url))
+            {
+                using (var response = await httpClient.SendAsync(request))
```
Comment (loic-sharma) [inline src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs]: Consider reducing nesting:

```cs
using (var request = new HttpRequestMessage(HttpMethod.Get, url))
using (var response = await httpClient.SendAsync(request))
{
   ...
}
```

Feel free to ignore if you prefer the current style

File: src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs (hunk 9 above)
Comment (loic-sharma) [reply src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs]: Does the `responseContent` contain owner information? If yes, do we need to worry about GDPR data privacy?

File: src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs (hunk 8 above)
Comment (loic-sharma) [reply src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs]: Could you add a comment about the `1 eq 1` trick? I didn't realize what this was doing until I found Joel's comment

Comment (cristinamanum) [reply src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs]: Updated for both comments.

File: src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs (hunk 9 above)
Comment (cristinamanum) [reply src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs]: Removed the log line. The data is the test data, it is not customer data.

File: src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs (hunk 11 above)
Comment (cristinamanum) [reply src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs]: Updated

File: src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs (hunk 10 above)
Comment (cristinamanum) [reply src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs]: The test is very simple I do not need the _httpClient.

File: src/NuGet.Services.EndToEnd/PushTests.cs (hunk 7 above)
Comment (cristinamanum) [reply src/NuGet.Services.EndToEnd/PushTests.cs]: Removed

File: src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs (hunk 10 above)
Comment (loic-sharma) [reply src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs]: I’m not sure I follow. What’re the benefits of not using `_httpClient`?

Comment (cristinamanum) [reply src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs]: Talked offline

**Pull request merged**